Use fix and extend function for all-packages.nix

nbp opened the pull request:
This pull request aims at making tiny incremental changes to `all-packages.nix` in order to progressively converge to what got achieved by @Mathnerd314 in #9400.

The goal of this pull request is to change the layout of the code without changing the data flow.  If this predicate is not honored, the patch includes a big comment describing how this is identical / different.

Note for reviewers:
- Review one patch at a time.
  - The entire set of changes might be hard to digest.
- Always leave a comment and a reaction (:+1:, :-1:, :confused:, …).
  - If you approve, disapprove, need clarification, or when you would be able to address this review.

cc @edolstra, @shlevy and @Mathnerd314

nbp committed:
Split miss-indented last lines. to simplify future diffs.

nbp committed:
Re-indent the top of pkgsFun.

nbp committed:
Move 'with helperFunctions' above the set of all packages.

nbp committed:
Move helperfunctions and stdenvAdapters under pkgsFun.

The `helperFunctions` and `stdenvAdapters` both use the `pkgs` attribute as
input, either to inherit some properties, either to use it as argument.

The `pkgs` binding used in both expressions of the `helperFunctions` and
`stdenvAdapters` is no longer the result of the `applyGlobalOverrides`
function, but the argument of the `pkgsFun` function.

The `pkgsFun` functions is called twice under `applyGlobalOverrides`, and in
both cases, the first argument of `pkgsFun` correspond to the result of
`applyGlobalOverrides`.

Thus, this modification will change the bindings, but the evaluation of
`<nixpkgs>`.

A third call the `pkgsFun` exists under `overridePackages` in the set of all
packages. Previously, the `helperFunctions` and `stdenvAdapaters` would use
the functions defined as part of the default `<nixpkgs>` set.  With this
modification, the `helperFunctions` and the `stdenvAdapters` are now using
the fix-point of the newly evaluated package set.

This implies that this modification allow the user to use
`overridePackages`, which is already not recommended for performance
reasons, to override the inputs of the `helperFucntions` and
`stdenvAdapaters` too, where this was not possible before.

nbp committed:
Move pkgsOrig outside applyGlobalOverrides and provide it as argument.

nbp committed:
Add an extra argument to the overrider function of applyGlobalOverrides.

nbp committed:
Unify the applyGlobalOverride function with the overridePackages function.

nbp committed:
Extract stdenvOverrides function out of mkOverrides function.

shlevy commented on pkgs/top-level/all-packages.nix:
```diff
@@ -16549,4 +16549,8 @@ tweakAlias = _n: alias: with lib;
-in lib.mapAttrs tweakAlias aliases // self; in pkgs
+in
+  lib.mapAttrs tweakAlias aliases // self;
+
```
Why this blank space?

nbp replied:
Because the lines above are the end of a function, while the 2 lines below are the end of an enclosing let-in which goes through the whole file.

As soon as all packages are moved in another file than the fix-point logic of nixpkgs, I will re-indent these lines properly.  In the mean time this extra line is a way to split the 2.

shlevy commented:
Overall seems fine for me.

nbp committed:
Move applyGlobalOverrides comment to stdenvOverrides, where it fits better.

nbp committed:
Move mkOverrides under applyGlobalOverrides.

nbp committed:
Under applyGlobalOverrides, inline mkOverrides in overrides.

nbp committed:
Remove with helperFunction above the list of all packages.

This modification change the names bound to the `helperFunctions` attribute
set, to be bound to `self` which is constructed by merging the same
`helperFunctions` set with the set of all packages.

This patch works as expected because none of the helperFunction names is
aliased by the name of a package.

nbp commented:
Note, in a similar way as #9400, this is blocking #10851 , as we need to be able to express nixpkgs as a single function to handle security issues properly.

nbp committed:
Under pkgsFun, move the merge of helper functions from self_ to self.

nbp committed:
Under pkgsFun, expand self attribute, and rename self_ to self.

Note, the aliases are now computed against the set of packages defined in
the set of all packages, and no longer apply to any overriden package.

I think this is better as this reduces the amount of surprizes.

nbp committed:
Split stdenv attribute set.

Extract stdenvDefault from the set of all packages.  As this set of
attributes are inter-dependant, probably due to stdenvOverrides, we have to
keep them in a close set of inter-dependent options.

I guess I will have to investigate more ...

zimbatm commented:
Just reviewed all the changes. In the end it's not a lot of lines that move but it was helpful to split it in different commits like that for the review.

nbp committed:
Swap conditions of stdenvOverrides to prevent infinite loops while evaluating stdenvCross.

While evaluating the derivation of xbursttools:
  the condition `pkgs.stdenv ? overrides` causes the evaluation of
  `stdenvCross`.  This evaluation comes too early during the execution, as
  it prevents the resolution of names such as `pkgs.lib`, and
  `stdenvAdapaters.makeStdenvCross`, which we want to take from `pkgs`
  instead of `self` in following patches.

By swapping the conditions, we effectively make the resolution of `pkgs.lib`
and `stdenvAdapaters.makeStdenvCross` possible through the pkgs attribute.

nbp committed:
Replace with statements of stdenvDefault to rely on pkgs instead of self.

nbp committed:
Move merge of stdenvDefault from the self attribute to the returned value of pkgsFun.

nbp committed:
Move tweakAlias function into the aliases.nix file, and rename it.

nbp committed:
Remove unused 'ncat' alias, as a package with the same name already exists.

nbp committed:
Move the set of aliases after 'self', such that we can later use the extend function.

nbp committed:
Move stdenvDefault into its own file.

nbp committed:
Move the merge of the overrides from pkgsFun to applyGlobalOverrides.

nbp committed:
Remove applyGlobalOverrides's pkgsOrig argument.

This patch change the behaviour or overridePackages:

 - Before, overridePackages was based on top of the set of packages
   overriden by `~/.nixpkgs/config.nix` file.

 - After, overridePackages should when given a function f, behave almost¹
   the same as-if the packageOverrides attribute of `~/.nixpkgs/config.nix`
   file attribute was set to f, assuming that we ignore the extra `self`
   argument.

¹ It is not yet exactly the same because we have many paths which are
re-entering nixpkgs while taking the default attributes, such as the
bootstrap phase of stdenv.  Thus if the nixpkgs configuration overrides any
of the dependencies need for the bootstrap, then we might get different sha.

nbp committed:
Within applyGlobalOverrides, alias pkgsFun calls.

nbp committed:
Rename applyGlobalOverrides to pkgsWithOverrides

nbp committed:
Reformat pkgsWithOverrides to later facilitate the identitication with the extend functions.

nbp committed:
Extract the top-level logic out of all-packages.nix into pkgs/top-level/default.nix

nbp committed:
Inline pkgsFun in pkgsWithOverrides.

nbp committed:
Split consecutive merge operators to simplify the identification of extend functions.

nbp committed:
Remove all-packages.nix helperFunctions dependency.

nbp committed:
Split helperFunctions to simplify the identification with the extend function.

nbp committed:
Move stdenvAdapaters attribute from all-packages.nix into the stdenvAdapters of the top-level.

nbp committed:
Build aliases based on the previous set including all-packages, instead of only using all-packages.

nbp committed:
Internalize the recursion over the set of all packages.

nbp committed:
In top-level/default.nix, rename self to allPackages.

nbp committed:
Add extra argument to better identity with the extend function later.

nbp committed:
Add bootStdenv condition around the overrider.

This condition used to be implicit in the sense that the stdenvOverrides
uses pkgs_6, instead of pkgs_7 as its super set, and also that each stage of
the compiler is build again the bootstrap packages of the previous stage,
thus overriden by the stdenv.override function.

nbp committed:
Move the customOverrides after the stdenvOverrides.

This change is not backward compatible, in the sense that this allow you to
override the packages which used to be overriden by the last stage of
stdenv.  Note, this does not mean that these packages would be used by
stdenv.

nbp committed:
Change the customOverrides to use the stdenvOverrides instead of the original package set.

nbp committed:
Under pkgsWithOverrides, replace the last uses of pkgs by the corresponding argument.

nbp committed:
Replace the merge operators by the fix' and extends functions.

nbp committed:
Move stdenvOverrides under pkgsWithOverrides.

nbp commented:
# Mission Complete

Waiting for reviews now.
Please try to address this as soon as possible to avoid letting these patches bit-rots.

As noted before, please review one patch at a time to follow the logic of the rewrites, otherwise, this changes might be harder to digest.

These changes are going a bit further than what is done in #9400, as these changes are also using the extend functions for replacing the `pkgsFun` content.  In addition, I also included the separation of the logic of `all-packages.nix` (#10307) and moved it into `top-level/default.nix`.  This makes this refactoring nicer.

zimbatm commented on pkgs/top-level/aliases.nix:
```diff
@@ -2,9 +2,21 @@ self:
 with self;
 
+let
+  # Removind recurseForDerivation prevents derivations of aliased attribute
```
typo: `Removind` -> `Removing`

nbp committed:
Fix comments typos.

zimbatm commented on pkgs/top-level/aliases.nix:
```diff
@@ -2,9 +2,21 @@ self:
+  removeRecurseForDerivations = _n: alias: with lib;
+    if alias.recurseForDerivations or false then
+      removeAttrs alias ["recurseForDerivations"]
+    else alias;
```
This function can be simplified to:
```nix
_n: alias: removeAttrs alias ["recurseForDerivations"]
```

nbp replied:
I thought of that, but I have not done it because this adds extra computations, which is not desirable.  For example if you have a large attribute set, you would decompose it to recompose it without one attribute.

zimbatm replied:
Shouldn't `removeAttrs` return the same set if nothing has changed ? In any case this is not new code so it can stay the same.

nbp replied:
https://github.com/NixOS/nix/blob/master/src/libexpr/primops.cc#L1060-L1080

Yes, the attribute will contain the same content, but this would be a new one.
I guess this would be easy to fix in Nix.

zimbatm commented on pkgs/top-level/default.nix:
```diff
@@ -0,0 +1,148 @@
+                lib.extends stdenvDefault (
+                  lib.extends trivialBuilders (
+                    lib.extends stdenvAdapters (
+                      self: {}))))))));
```
I guess that's going to be the controversial bit. Is there any benchmark we can do for before / after ?

zimbatm replied:
I really like how all the previous commits where leading to this change.

vcunat replied:
Somehow, I dislike deep nesting. Since we don't have the function-composition operator (yet), we might go for a fold instead, which seems perhaps even more natural here. I believe this is equivalent (tested some evals):
```nix
let in # to fix highlighting
      with lib; fix' (fold extends (self: {})
        [ customOverrides stdenvOverrides aliases allPackages
          stdenvDefault trivialBuilders stdenvAdapters
        ])
```

Ericson2314 replied:
Super-tiny-I-feel-even-pointing-this-out-beta-expansion-nit: I'd do
```nix
let
  extenders = with lib; [
    customOverrides stdenvOverrides aliases allPackages
    stdenvDefault trivialBuilders stdenvAdapters
  ];
in fix' (fold extends (self: {}) extenders)
```

nbp replied:
This indeed looks shorter, and this removes the bunch parenthesis which might let people believe that Scheme is better.

Still one complain I have with the current, and the suggested change, is that the list is in reverse order.  We start reading from the latest applied to the first applied.  Thus I would prefer us to use:

Also, I would prefer if we were to put only one function per line, as these are quite complex, and the important aspect of the code is the sequencing, not the collection.

```
let
  layers = [
    stdenvAdapaters
    trivialBuilders
    stdenvDefault
    allPackages
    aliases
    stdenvOverrides
    customOverrides
  ];
in
  foldl (flip extends) (self: {}) layers
```

Ericson2314 replied:
Additionally with 1/line, shorter ones like `aliases` and `stdenvOverrides` could perhaps be inlined.

vcunat replied:
Yes, one per line is probably better, even though it's just collecting a list of bindings let-defined above. I'm irresolute about the order; in any case, we might best add a comment to signify which end is the innermost and which is the outermost, and we should better keep the order consistent with the order in which we define the extending functions.

You propose starting with the innermost layer, which would be the order if evaluated strictly. It might be more natural to most people. Note however that we evaluate lazily, so the real order "starts" evaluating with the outermost layer. (The innermost layer "finishes" first, of course.)

Ericson2314 replied:
Don't care on inner/outer, +1 comment explaining regardless.

Profpatsch replied:
> we might best add a comment to signify which end is the innermost and which is the outermost

I really like how expressive `foldl` is: “starts from the left and finishes on the right”. But of course, document whatever is not trivial.

zimbatm commented:
I need to make another pass tomorrow. Most of it is moving code around which is not too controversial and could be merged quickly.

zimbatm commented on pkgs/top-level/all-packages.nix:
```diff
@@ -97,6 +97,10 @@ let
+  #
+  # We don't want stdenv overrides in the case of cross-building, or
+  # otherwise the basic overrided packages will not be built with the
+  # crossStdenv adapter.
```
This change could be squashed with the previous one

zimbatm commented on pkgs/top-level/all-packages.nix:
```diff
@@ -83,6 +83,11 @@ let
+  topLevelArguments = {
+    inherit system bootStdenv noSysDirs gccWithCC gccWithProfiling config
+      crossSystem platform lib;
+  };
```
Aliasing these arguments seem a bit superfluous if they can be passed to `import ./stdenv {}` directly. A later change might invalidate my thinking.

vcunat replied:
Later it's used on two places.

zimbatm commented on pkgs/top-level/default.nix:
```diff
@@ -124,7 +124,9 @@ let
-      self = (import ./all-packages.nix selfArgs) self pkgs;
+      self =
+        let res = import ./all-packages.nix selfArgs res pkgs;
+        in res;
```
This change could be part of 01693ac154138e95324959d260056312021e0dae

zimbatm commented on pkgs/top-level/default.nix:
```diff
@@ -113,33 +113,34 @@ let
 
       # The overriden, final packages.
-      pkgs =   pkgs_7 // stdenvOverrides pkgs_6;
+      pkgs =   pkgs_7 // stdenvOverrides pkgs pkgs_6;
```
I guess this is the commit that starts introducing a bit of overhead with additional function calls ?

nbp replied:
No, this commit should have no overhead as the arguments are not used, which implies that the argument are not needed as soon as the function is applied.

zimbatm commented:
Finished reviewing all the changes. Once we agree on the changes most of this should be squashed into bigger commits don't you think ?

Overall +1 for separating the all-packages in more logical chunks and making everything more flexible. I would like some more testing to make sure that cross-compilation, and `~/.nixpkgs/config.nix`. I also don't know what performance impact these changes are going to be. If that's the case then we should split the changes into two so that at least the first part can go in.

nbp commented:
> Once we agree on the changes most of this should be squashed into bigger commits don't you think ?

I do not think there is much value in squashing the commits, as once merged we would have a single commit in nixpkgs.

This would also be bad if we ever want to bisect through the changes, since this might help us investigate the issue better.

nbp commented:
> I would like some more testing to make sure that cross-compilation

In fact, when running nox-review, we are already generating a package which uses a cross compiler, and do that by setting crossSystem.  So this is actually tested by nox-review.

> and `~/.nixpkgs/config.nix`

I already a non-empty config.nix, so I will double that these are getting the same sha.

For the override of stdenv packages, I honestly do not see much value, except adding debug symbols to them and I would be fine to restore the old behaviour back, if requested.

aszlig commented on pkgs/top-level/default.nix:
```diff
@@ -0,0 +1,148 @@
+      # mechanism to implement some staged compilation of the stdenv.
+      #
+      # We don't want stdenv overrides in the case of cross-building, or
+      # otherwise the basic overrided packages will not be built with the
```
Spelling: `overridden`

aszlig commented on pkgs/top-level/default.nix:
```diff
@@ -0,0 +1,148 @@
+  # The `packageOverrides' function is called with the *original*
+  # (un-overriden) set of packages, allowing packageOverrides
+  # attributes to refer to the original attributes (e.g. "foo =
+  # ... pkgs.foo ...").
```
`s/overriden/overridden/g`

zimbatm commented:
When comparing `time nix-env -f. -qaP > /dev/null` this branch is a tiny bit faster that the commit it's based over. Over multiple runs I get approximately:

Before:
```
real	0m2.094s
user	0m1.908s
sys	0m0.185s
```

After:
```
real	0m2.070s
user	0m1.915s
sys	0m0.153s
```

Ericson2314 commented:
I've been a major fan of these changes through many iterations, but the way this is split up is just beautiful. I will recommend this PR as a followup to anyone wishing to learn more about how nixpkgs is structured after reading the nix pills.

nbp commented:
Apparently the split of stdenv seems to have cause issues with `stdenv_32bits`, when used in overriden packages, as well as the split of `stdenvOverrides` and `customOverrides`.  I am investigating the `drv` files differences to see from where this issue comes from. (wish nix-store would output diff-able results)

nbp commented:
Ok, I will check that later, but my guess is that when I separated the stdenv to use `with pkgs` instead of `with self` (commit 020bb40), I left over a few bindings which are unfortunately redefined by `stdenvOverrides`.

Thus my guess is that the `gcc` variable of `gcc_multi`[1] is now bound to the `stage4` [2], instead of the all-packages one [3].

[1] https://github.com/nbp/nixpkgs-2/blob/fix-extend/pkgs/top-level/all-packages.nix#L3976
[2] https://github.com/NixOS/nixpkgs/blob/master/pkgs/stdenv/linux/default.nix#L304
[3] https://github.com/nbp/nixpkgs-2/blob/fix-extend/pkgs/top-level/all-packages.nix#L4087

nbp commented:
Ok, I cannot explain from where the difference of evaluation is coming from, but now I tempted to not fix it at all.

The reason why I will not fix it is because now, if we write the same expression in `config.nix`, as we were to write it in `all-packages.nix`, then we would get the same hashes.  Which was not the case before.

I think this is a non-neglectable addition for landing this fix-extend approach.
I am just sad I cannot explain why the old version is using `fetchurlBoot` sooner than expected.

zimbatm commented:
It's not clear to me what the issue is. Does it cause an error or just changes the derivation hash ? What do you do to reproduce the issue ?

peti commented:
@nbp, thank you very much for the effort you've put into this clean-up and for submitting the change as a series of easily reviewable patches. I'd would be great to get those changes merged into `master` ASAP. Even *if* this re-factoring does have unintended consequences that we find out only after merging it, then I'm confident that we can find a way to fix them, so IMHO the risk of merging this update is rather small.

nbp commented:
Thanks @zimbatm , @aszlig and @vcunat for reporting the typos while reviewing the commits :)

@zimbatm 
> It's not clear to me what the issue is. Does it cause an error or just changes the derivation hash ? What do you do to reproduce the issue ?

I tested to clone `rr` expression in my `config.nix` file, and renaming it `rr4` to make sure I do not get the old one, and checked the hash of the derivation with `nix-instantiate ./ -A rr4`.  I was able to evaluate both fine, but got different hashes.

Then I compared the hash with the hash of the derivation of `nix-instantiate ./ -A rr`, before and after this patch series. I got a different result for the base branch, and an identical result of this branch.

This is why I mentioned that I am not going to fix this issue, as I get the same behaviour in `config.nix` as-if the package was written in `all-packages.nix`.

```
base-branch$ nix-instantiate ./ -A rr4
/nix/store/gbg1jpm6nb8z2kcgjm0qw8ysy56z0a6j-rr-4.2.0.drv
base-branch$ nix-instantiate ./ -A rr
/nix/store/y7xqd9i08f09nhqz8ypfhv4jcy6dr1ym-rr-4.2.0.drv
fix-extend$ nix-instantiate ./ -A rr4
/nix/store/y7xqd9i08f09nhqz8ypfhv4jcy6dr1ym-rr-4.2.0.drv
fix-extend$ nix-instantiate ./ -A rr
/nix/store/y7xqd9i08f09nhqz8ypfhv4jcy6dr1ym-rr-4.2.0.drv
```

vcunat commented:
Thanks a lot for cleaning the override mess and making the transition reviewable! (I finished reviewing this a bit late.) In some commits I didn't really see the motivation why the new state should be preferable to the old one, but on the whole it seems much easier to understand now.

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1,133 +1,16 @@
-
-
-let config_ = config; platform_ = platform; in # rename the function arguments
+with pkgs;
```
Should this be `self`? `pkgs` = `super`, right? I am reading https://github.com/NixOS/nixpkgs/pull/10851/commits/295b7a682b6b2c1bd8f9922b8721137fd93921c8 and getting even more confused :).

nbp replied:
In the final result, self is equivalent to `rec`, and `pkgs` is equivalent to the `self` of the fix-point.  `super` would be all the `stdenvAdapaters` and the `trivialBuilders`.

Ericson2314 replied:
Ah, I missed https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/default.nix#L119

Ericson2314 replied:
`grep --color -E 'self( |$|\.)' pkgs/top-level/all-packages.nix` Convinces me we could remove the `self` arg altogether from all-packages and put everything in aliases.

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -1,133 +1,16 @@
+with pkgs;
 
 let
+  defaultScope = pkgs // pkgs.xorg;
```
Likewise right now shouldn't `defaultScope` be made from self?

nbp replied:
As `self` corresponds to `rec` at the moment, we should not do that as this will break the `packageOverride` mechanism, and the security update branch.

Ericson2314 replied:
Ok based on other thread it is the fixed point being used here as I wanted all along :).

Ericson2314 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -157,11 +40,7 @@ let
-      newpkgs = pkgsFun newpkgs overrides;
-      overrides = mkOverrides pkgs (f newpkgs pkgs);
-    in newpkgs;
+  overridePackages = f: pkgsWithOverrides f;
```
This replaces `packageOverrides` in config with something else. but sometimes one wants to chain overrides. I'd like to see a
```nix
let # for syntax coloring
  reoverridePackages = f: fix' (extend f self.__unfix__);
```

nbp replied:
I am not sure what should be done with this `overridePackages` function.  I personally have no uses of it, and I am sure we can discuss that, or just remove it if not used.

I did not went for using the unfix attribute for one simple reason.  The current version of `overridePackages` does not contain any of the modifications made in `~/.nixpkgs.config.nix`.  Thus, adding such a thing would probably be bad, as one might rely on such feature to provide its own set of overridden packages inside a repository.

Ericson2314 replied:
Fair enough, this can be dealt with separately. The fact that nixpkgs even has such `~/.nixpkgs.config.nix` etc hardcodings kinda scares me, but at least it's cordoned from all-packages now.

nbp commented:
@vcunat 
> so this should finally remove the distinction between using foo and pkgs.foo in the inside of the file :-)

Mainly, yes, but with the security branch which is coming we should default to `pkgs.foo` for dependencies, and default to `self.foo` for statically linked packages, or for overridden packages.

> Maybe it would be useful to have a variant of the // operator that would throw an error on collision instead of overriding...

I guess we could add a check for that as part of the `aliases.nix` file.
